frontend/appflowy_tauri/src/appflowy_app/components/_shared/EditRow/CheckListPopup.tsx

```diff
@@ -0,0 +1,179 @@
+import { CheckListProgress } from '$app/components/_shared/CheckListProgress';
+
+export const CheckListPopup = ({
+  left,
+  top,
+  cellIdentifier,
+  cellCache,
+  fieldController,
+  openCheckListDetail,
+  onOutsideClick,
+}: {
+  left: number;
+  top: number;
+  cellIdentifier: CellIdentifier;
+  cellCache: CellCache;
+  fieldController: FieldController;
+  openCheckListDetail: (left: number, top: number, option: SelectOptionPB) => void;
+  onOutsideClick: () => void;
+}) => {
+  const databaseStore = useAppSelector((state) => state.database);
+  const { data } = useCell(cellIdentifier, cellCache, fieldController);
+  const { t } = useTranslation('');
+
+  const [allOptionsCount, setAllOptionsCount] = useState(0);
+  const [selectedOptionsCount, setSelectedOptionsCount] = useState(0);
+  const [newOptions, setNewOptions] = useState<string[]>([]);
+
+  useEffect(() => {
+    setAllOptionsCount(
+      (databaseStore.fields[cellIdentifier.fieldId]?.fieldOptions as ISelectOptionType)?.selectOptions?.length || 0
+    );
+  }, [databaseStore, cellIdentifier]);
+
+  useEffect(() => {
+    setSelectedOptionsCount((data as SelectOptionCellDataPB)?.select_options?.length || 0);
+  }, [data]);
+
+  const onToggleOptionClick = async (option: SelectOptionPB) => {
+    if ((data as SelectOptionCellDataPB)?.select_options?.find((selectedOption) => selectedOption.id === option.id)) {
+      await new SelectOptionCellBackendService(cellIdentifier).unselectOption([option.id]);
+    } else {
+      await new SelectOptionCellBackendService(cellIdentifier).selectOption([option.id]);
+    }
+  };
+
+  const onCheckListDetailClick = (e: any, option: ISelectOption) => {
+    e.stopPropagation();
+    let target = e.target as HTMLElement;
+
+    while (!(target instanceof HTMLButtonElement)) {
+      if (target.parentElement === null) return;
+      target = target.parentElement;
+    }
+
+    const selectOption = new SelectOptionPB({
+      id: option.selectOptionId,
+      name: option.title,
+      color: option.color || SelectOptionColorPB.Purple,
+    });
+
+    const { right: _left, top: _top } = target.getBoundingClientRect();
+    openCheckListDetail(_left, _top, selectOption);
+  };
+
+  const newOptionClick = () => {
+    setNewOptions([...newOptions, '']);
+  };
+
+  const updateNewOption = (index: number, value: string) => {
+    const newOptionsCopy = [...newOptions];
+    newOptionsCopy[index] = value;
+    setNewOptions(newOptionsCopy);
+  };
+
+  const onNewOptionKeyDown = (index: number, e: KeyboardEvent) => {
+    if (e.key === 'Enter') {
+      void onSaveNewOptionClick(index);
+    }
+  };
+
+  const onSaveNewOptionClick = async (index: number) => {
+    await new SelectOptionCellBackendService(cellIdentifier).createOption({ name: newOptions[index] });
+    setNewOptions(newOptions.filter((_, i) => i !== index));
+  };
+
+  return (
+    <PopupWindow className={'text-xs'} onOutsideClick={onOutsideClick} left={left} top={top}>
```

Comment:
Sometimes I feel that too long element codes will confuse me because I can’t directly map them into corresponding components in my mind, so I prefer to name and abstract different elements

Reply:
I generally agree, but I can't tell how I can improve this code. Do you have any suggestions?

Reply:
Ok. I added some component names, but you can add more correct names according to the actual situation